url47ebb27197_asirbu_CorrectiveEvolution_tgz-pJ8-eu_fbk_soa_evolution_engine_impl_test_ProblemToSMVFuncTestJ8/src/eu/fbk/soa/evolution/engine/impl/STSToSMV.java

```diff
@@ -0,0 +1,356 @@
+		String translation = "";
+		
+		Set<Action> actions = new HashSet<Action>(totalActions);
+		for (Action act : totalActions) {
+			if (!act.getRelatedEntity().equals(translatedEntity)) {
+				actions.remove(act);
+			}
+		}
+		if (!actions.isEmpty()) { 
+			translation += "IVAR " + type + ": {UNDEF";
+			Set<Action> noUndefActions = new HashSet<Action>(totalActions);
+			for (Action act : totalActions) {
+				if (act.getName().equals("UNDEF")) {
+					noUndefActions.remove(act);
+				}
+			}
+			String actionsDecl = StringUtils.getCommaSeparatedString(noUndefActions);
+			if (actionsDecl != "") {
+				translation += ", " + actionsDecl;
+			}
+			translation += "};\n";
+		}
+		return translation;
+	}
+	
+	private String translateTransitionGroup(Set<Transition> trGroup,
+			Object translatedEntity) {
+		
+		String translation = "";
+		if (trGroup.size() >= 1) {
+			Transition trans = trGroup.iterator().next();
+			
+			translation +=
+				this.translateTransitionCondition(trans, translatedEntity);
+			translation += "& " + translateAction(trans.getAction(), translatedEntity) + " ";	
+			
+			translation += ": ";
+			
+			int nr = 0;
+			String targetStates = "";
+			for (Transition t : trGroup) {
+				State target = t.getTarget();
+				if (nr > 0) {
+					targetStates += ", ";
+				}
+				targetStates += target.getName();
+				nr++;
+			}
+			if (nr == 1) {
+				translation += targetStates;
+			} else {
+				translation += "{" + targetStates + "}";
+			}
+			
+			translation += ";\n";
+		}
+		return translation;
+	}
+	
+	private String translateTransition(Transition trans, Object translatedEntity) {
+		String translation = 
+			this.translateTransitionCondition(trans, translatedEntity);
+		
+		translation += "& " + translateAction(trans.getAction(), translatedEntity) + " ";	
+		
+		translation += ": " + trans.getTarget() + ";\n";
+		return translation;	
+	}
+	
+	private String translateTransitionCondition(Transition trans, Object translatedEntity) {
+		String translation = "(state = " + trans.getSource() + ") ";
+		
+		String formula = translateStateFormula(trans.getCondition());
+		if (formula != "") {
+			translation += "& " + formula;
+		}
+		
+		for (Clause<? extends Literal> clause : trans.getGuardClauses()) {
+			String clauseCond = "";
+			Iterator<? extends Literal> iterator = clause.getLiterals().iterator();
+			
+			while (iterator.hasNext()) {
+				Literal lit = iterator.next();
+				String guardCond = "";
+				if (lit.isNegated()) {
+					guardCond += "!(";
+				}
+				for (String entityID : id2STS.keySet()) {
+					STS sts = id2STS.get(entityID);
+
+					Set<State> states = sts.getStatesForLabel(lit.getProposition());
+					for (State state : states) {
+						if (guardCond != "" && !guardCond.equals("!(")) {
+							guardCond += " | ";
+						}
+						guardCond += entityID + ".state = " + state.getName();
+					}
+				}
+				if (lit.isNegated()) {
+					guardCond += ")";
+				}
+				if (guardCond != "") {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.